### PR TITLE
Integrate NeonVM into Autoscaling

### PR DESCRIPTION
**Pull request by bayandin**
General comment:
I've cut some corners while merging these projects; the main goal was to get all the parts to play well in the single repo without introducing unnecessary refactoring. Thus some parts look a bit rough and could be improved (for example, targets in Makefile); a difference in the project structure and directory layout is still visible. Nevertheless, I still see the benefits of getting this merged and continuing to work on both NeonVM and Autoscaling in the same repo.

ToDo list mostly for myself:

- [x] Merge repos https://github.com/neondatabase/autoscaling/pull/96
- [x] Fix make targets
- [x] Fix tests
- [x] Fix e2e tests
- [x] Unify .github/workflows
- [x] Fix scripts (for autoscaling)
- [x] sort out docs / READMEs
- [x] restore git history of neonvm https://github.com/neondatabase/autoscaling/pull/96
- [ ] Fix `golangci-lint` for neonvm (can be done in a separate PR)
- [ ] Test release workflow (create `0.5.0-pre` tag?)
- [x] Move tuning `fs.inotify.*` settings from GitHub job to ansible config https://github.com/neondatabase/runner/pull/20

### CODE REVIEW

File: .github/workflows/release.yaml
Comment (bayandin) [inline .github/workflows/release.yaml:51]: In NeonVM we have this code in [`release.yaml`](https://github.com/neondatabase/neonvm/blob/d9ce98c5d7e949ff25049d9b6dab7f2142184290/.github/workflows/release.yaml#L34-L39), but I far as I understand we commit all autogenerated code to the repo, so this step shouldn't be necessary, am I missing something?
/cc @cicdteam

File: Makefile
Comment (sharnoff) [inline Makefile:13]: Is `DIST` instead of `DISK` a typo?

File: kind/config.yaml
Comment (sharnoff) [inline kind/config.yaml:5]:
```suggestion
```

File: neonvm/hack/Dockerfile.generate
Comment (sharnoff) [inline neonvm/hack/Dockerfile.generate:8]: This seems funky - why is this needed? (or: can we have a comment explaining the purpose of this?)

I'm particularly wary of this because I have to run docker stuff as root, so this _looks_ like it might mess things up, but I can't really tell either way.

File: tests/e2e/autoscaling/00-create-vm.yaml
Comment (sharnoff) [inline tests/e2e/autoscaling/00-create-vm.yaml:15]: Is this related?

**Review by sharnoff — APPROVED** [review]
Broadly looks good. The comments are mostly little things.

A couple notes:

1. A few of the changes don't seem strictly related to the PR — it'd be good to separate those out, to avoid giving the impression to future `git blame`s that they were related/required here.
2. I noticed a couple references to `vm_image` still around. Aside from the one I commented on, there's also references in `README.md` and `vm-examples/pg14-disk-test/README.md`

I'm a little concerned about some of the `kubectl` usage in the makefile, but AFAICT it's not meaningfully different from what we had before, so I don't think that should block merging. I might open up a PR to improve it sometime down the road.

---

Lastly: It seems like you're working on macos — have you already tested that things work as expected on linux as well? would it be useful for me to do that?

File: neonvm/hack/Dockerfile.generate
Comment (bayandin) [reply neonvm/hack/Dockerfile.generate:8]: I've cloned the repo as a local non-root user. Then if I run `make generate` (which uses this file), it generates files in a container from the root user. And I have issues deleting these files (during `git rebase`, for example, because they don't belong to the current user), so I've changed it to use the current user.

There's a comment regarding this in `Makefile` (for `generate` target), will add it to the Dockerfile as well.
```
# Use uid and gid of current user to avoid mismatched permissions
```

> I'm particularly wary of this because I have to run docker stuff as root, so this looks like it might mess things up, but I can't really tell either way.

I guess it just wouldn't work for your case 😄 (because the root user already exists in a container). I will add a fix for it.

**Review by bayandin — COMMENTED** [review]
Thanks for the review @sharnoff!


> A few of the changes don't seem strictly related to the PR — it'd be good to separate those out, to avoid giving the impression to future git blames that they were related/required here.

Before merging PR, I'm going to squash commits into a couple of more meaningful ones (like `fix e2e tests`) and rebase on top of the `main` branch (without squashing everything into one commit). Hope this will help to make `git blame` more meaningful.

> I noticed a couple references to vm_image still around. Aside from the one I commented on, there's also references in README.md and vm-examples/pg14-disk-test/README.md

My bad, I will check and update these.

> Also, I think it's best to fix golangci-lint in a separate PR.

Sure! 

> Lastly: It seems like you're working on macos — have you already tested that things work as expected on linux as well? would it be useful for me to do that?

I'm using an ec2 Linux instance for this, `neonvm` can't be compiled yet on macOS because of a difference in netlink API (@cicdteam have some fixes for this in his [neonvm `arm64` branch](https://github.com/neondatabase/neonvm/compare/main...arm64))
But yes, it works on Linux for me

File: neonvm/controllers/virtualmachine_controller_test.go
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller_test.go:84]: Makes sense, sounds good

File: neonvm/hack/Dockerfile.generate
Comment (sharnoff) [reply neonvm/hack/Dockerfile.generate]: Makes sense - validated the fix works :)

File: tests/e2e/autoscaling/00-create-vm.yaml
Comment (sharnoff) [reply tests/e2e/autoscaling/00-create-vm.yaml:15]: Makes sense

**Comment by sharnoff** [comment]
> > A few of the changes don't seem strictly related to the PR — it'd be good to separate those out, to avoid giving the impression to future git blames that they were related/required here.
> 
> Before merging PR, I'm going to squash commits into a couple of more meaningful ones (like fix e2e tests) and rebase on top of the main branch (without squashing everything into one commit). Hope this will help to make git blame more meaningful.

Sounds good! Yeah, seems like most of what I thought was unrelated is more ambiguous - my bad.

> I'm using an ec2 Linux instance for this, neonvm can't be compiled yet on macOS because of a difference in netlink API

Ah shoot, interesting. Out of curiosity, would compiling neonvm in containers (like what we do for the scheduler & autoscaler-agent) also fix that?

File: .github/workflows/release.yaml
Comment (cicdteam) [reply .github/workflows/release.yaml:51]: seems yes, we can skip code generating in GA

**Review by cicdteam — APPROVED** [review]

**Comment by bayandin** [comment]
> Ah shoot, interesting. Out of curiosity, would compiling neonvm in containers (like what we do for the scheduler & autoscaler-agent) also fix that?

Yep, it compiles (but we don't have arm64 kernel for `neonvm` yet, so can't say if it really works). I'd like to move the building of everything to containers, then we can improve caches (probably) and not require golang on host machine

**Comment by cicdteam** [comment]
> Yep, it compiles (but we don't have arm64 kernel for `neonvm` yet, so can't say if it really works). I'd like to move the building of everything to containers, then we can improve caches (probably) and not require golang on host machine

Yes, netlink has no issue when cimplining in container (because there us Linux, not Darwin). I able run neonvm on my Apple M1 macbook (with `hvm` acceleration ever). I will add arm64 support after we finish moving neonvm into autiscaling repo.

**Pull request merged**
